src/lib/components/availability/SetAvailabilityModal.svelte

```diff
@@ -0,0 +1,170 @@
+<script lang="ts">
+	import { invalidateAll } from '$app/navigation'
+	import { db } from '$lib/firebase/client/firebase'
+	import type { Availability, NewAvailability } from '$lib/types/domain/availability'
+	import type { Rehearsal } from '$lib/types/domain/rehearsal'
+	import { newTimeOnDay } from '$lib/util/date'
+	import { getTimeString } from '$lib/util/timeString'
+	import {
+		Modal,
+		Grid,
+		Row,
+		Column,
+		TextInput,
+		Checkbox,
+		TimePicker
+	} from 'carbon-components-svelte'
+	import { Edit } from 'carbon-icons-svelte'
+	import { collection, doc, setDoc, Timestamp } from 'firebase/firestore'
+
+	export let rehearsal: Rehearsal
+	export let user: { id: string; name: string }
+	export let availability: Availability | undefined = undefined
+
+	export let displayUserName: boolean = false
+
+	export let openSetAvailability = false
+	export let onClose: () => void
+
+	let setOrUpdate = availability ? 'Update' : 'Set'
+
+	let available = availability?.available ?? true
+	let remark: string | undefined = availability?.reason
+
+	let startTimeString: string = getTimeString(availability?.startTime ?? rehearsal.startTime)
+	let endTimeString: string = getTimeString(availability?.endTime ?? rehearsal.endTime)
+	let startTimeInvalid: boolean = false
+	let endTimeInvalid: boolean = false
+
+	/**
+	 * Submit or update the availability in the database.
+	 * @returns `true` if the mutation was successful, `false` otherwise.
+	 */
+	async function saveAvailability(): Promise<boolean> {
+		// Time input validation
+		const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/
+		startTimeInvalid = !timeRegex.test(startTimeString)
+		endTimeInvalid = !timeRegex.test(endTimeString)
+		if (startTimeInvalid || endTimeInvalid) return false
```

Comment:
Is probably used elsewhere, seperate and reuse

Reply:
It should be used somewhere else (rehearsal creation), but actually is not. I did abstract it though to re-use in a later issue